storage/storage_reference.go

```diff
@@ -0,0 +1,96 @@
+package storage
+
+import (
+	"github.com/Sirupsen/logrus"
+	"github.com/containers/image/docker/reference"
+	"github.com/containers/image/types"
+	"github.com/containers/storage/storage"
+)
+
+// A storageReference holds a name and/or an ID, which is a 32-byte value
+// hex-encoded into a 64-character string.
+type storageReference struct {
+	store     storage.Store
+	transport *storageTransport
+	reference string
+	id        string
+}
+
+func newReference(store storage.Store, transport *storageTransport, reference, id string) *storageReference {
+	return &storageReference{
+		store:     store,
+		transport: transport,
+		reference: reference,
+		id:        id,
+	}
+}
+
+// Resolve the reference's name to an image ID in the storage library if
+// there's already one present with the same name or ID.
+func (s *storageReference) ID() string {
+	if s.id == "" {
+		image, err := s.store.GetImage(s.reference)
+		if image != nil && err == nil {
+			s.id = image.ID
+		}
+	}
+	return s.id
+}
+
+func (s *storageReference) Transport() types.ImageTransport {
+	return s.transport
+}
+
+func (s *storageReference) DockerReference() reference.Named {
+	if s.reference == "" {
+		return nil
+	}
+	name, err := reference.ParseNamed(s.reference)
+	if err != nil {
+		return nil
+	}
+	return reference.WithDefaultTag(name)
+}
+
+func (s *storageReference) StringWithinTransport() string {
+	if s.reference == "" {
+		return s.id
+	}
+	return s.reference
+}
+
+func (s *storageReference) PolicyConfigurationIdentity() string {
+	if s.reference == "" {
+		return s.id
+	}
+	return s.reference
+}
+
+func (s *storageReference) PolicyConfigurationNamespaces() []string {
+	return nil
+}
+
+func (s *storageReference) NewImage(ctx *types.SystemContext) (types.Image, error) {
+	return newImage(s)
```

Comment:
All of the `s.Generic` forwarding etc. seems unnecessary, is there any benefit to doing it this way? AFAICS the only thing this does differently is that `Manifest()` and `Signatures()` override the caching done in `image.sourcedImage`.

It seems simpler to call ~`image.FromSource(newImageSource)` in `storageReference.NewImage` and drop the `types.Image` implementation from here.